File: rust/src/transport/in_process.rs
// SPDX-License-Identifier: MIT

use super::TransportService;
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

/// In-memory transport for testing and demos
/// Simulates a mesh network using a static registry
pub struct InProcessTransport {
    local_uhid: String,
    network: Arc<Mutex<HashMap<String, Arc<Mutex<Vec<u8>>>>>>,
    data_received_handler: Option<Box<dyn Fn(&str, &[u8]) + Send + Sync>>,
}

impl InProcessTransport {
    /// Creates a new in-process transport node
    pub fn new(local_uhid: String) -> Self {
        InProcessTransport {
            local_uhid,
            network: Arc::new(Mutex::new(HashMap::new())),
            data_received_handler: None,
        }
    }

    /// Registers this node in the simulated network
    pub fn register(&self) -> Result<(), Box<dyn std::error::Error>> {
        let mut net = self.network.lock().map_err(|e| format!("{}", e))?;
        if net.contains_key(&self.local_uhid) {
            return Err("Node already registered".into());
        }
        net.insert(self.local_uhid.clone(), Arc::new(Mutex::new(Vec::new())));
        Ok(())
    }

    /// Unregisters this node from the simulated network
    pub fn unregister(&self) -> Result<(), Box<dyn std::error::Error>> {
        let mut net = self.network.lock().map_err(|e| format!("{}", e))?;
        net.remove(&self.local_uhid);
        Ok(())
    }
}

#[async_trait]
impl TransportService for InProcessTransport {
    fn name(&self) -> &str {
        "InProcess"
    }

    fn is_available(&self) -> bool {
        true
    }

    fn max_bandwidth_bps(&self) -> i64 {
        1_000_000_000 // 1 Gbps
    }

    fn max_range_meters(&self) -> i32 {
        0 // Not applicable for in-process
    }

    fn power_cost_relative(&self) -> i32 {
        0
    }

    fn max_concurrent_peers(&self) -> i32 {
        i32::MAX
    }

    async fn send_async(&self, peer_uhid: &str, data: &[u8]) -> Result<bool, Box<dyn std::error::Error>> {
        if peer_uhid.is_empty() {
            return Ok(false);
        }

        let net = self.network.lock().map_err(|e| format!("{}", e))?;

        if let Some(peer_buffer) = net.get(peer_uhid) {
            let mut buf = peer_buffer.lock().map_err(|e| format!("{}", e))?;
            buf.extend_from_slice(data);
            Ok(true)
        } else {
            Ok(false)
        }
    }

    async fn send_stream_async(
        &self,
        peer_uhid: &str,
        stream: &mut (dyn std::io::Read + Send + Unpin),
    ) -> Result<bool, Box<dyn std::error::Error>> {
        use std::io::Read;

        let mut data = Vec::new();
        stream.read_to_end(&mut data)?;
        self.send_async(peer_uhid, &data).await
    }

    fn is_connected(&self, peer_uhid: &str) -> bool {
        if let Ok(net) = self.network.lock() {
            net.contains_key(peer_uhid)
        } else {
            false
        }
    }

    fn set_data_received_handler(
        &mut self,
        handler: Box<dyn Fn(&str, &[u8]) + Send + Sync>,
    ) {
        self.data_received_handler = Some(handler);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_send_between_nodes() {
        let mut node_a = InProcessTransport::new("node-a".to_string());
        let mut node_b = InProcessTransport::new("node-b".to_string());

        let network = Arc::new(Mutex::new(HashMap::new()));
        node_a.network = network.clone();
        node_b.network = network.clone();

        node_a.register().unwrap();
        node_b.register().unwrap();

        let data = b"test message";
        let result = node_a.send_async("node-b", data).await.unwrap();
        assert!(result);
    }

    #[tokio::test]
    async fn test_send_to_nonexistent_node() {
        let mut node_a = InProcessTransport::new("node-a".to_string());
        node_a.register().unwrap();

        let data = b"test message";
        let result = node_a.send_async("nonexistent", data).await.unwrap();
        assert!(!result);
    }

    #[test]
    fn test_is_connected() {
        let mut node_a = InProcessTransport::new("node-a".to_string());
        let mut node_b = InProcessTransport::new("node-b".to_string());

        let network = Arc::new(Mutex::new(HashMap::new()));
        node_a.network = network.clone();
        node_b.network = network.clone();

        node_a.register().unwrap();
        node_b.register().unwrap();

        assert!(node_a.is_connected("node-b"));
        assert!(!node_a.is_connected("nonexistent"));
    }
}

File: go/transport/in_process.go
// SPDX-License-Identifier: MIT

package transport

import (
	"context"
	"fmt"
	"sync"
)

// InProcessTransport is an in-memory transport for testing and inter-process communication.
// It uses a global sync.Map to route packets between nodes.
type InProcessTransport struct {
	name            string
	available       bool
	maxBandwidth    int64
	maxRange        int32
	powerCost       int32
	maxConcurrency  int32
	connectedPeers  sync.Map // map[string]bool
	messageHandlers sync.Map // map[string]chan []byte
	m               *PerTransportMetrics
}

// NewInProcessTransport creates a new in-memory transport.
func NewInProcessTransport() *InProcessTransport {
	return &InProcessTransport{
		name:           "InProcess",
		available:      true,
		maxBandwidth:   1000000, // 1 Mbps equivalent
		maxRange:       100,
		powerCost:      1,
		maxConcurrency: 100,
		m:              NewPerTransportMetrics(),
	}
}

// Name returns the transport name.
func (ipt *InProcessTransport) Name() string {
	return ipt.name
}

// IsAvailable returns whether the transport is available.
func (ipt *InProcessTransport) IsAvailable() bool {
	return ipt.available
}

// MaxBandwidthBps returns maximum bandwidth.
func (ipt *InProcessTransport) MaxBandwidthBps() int64 {
	return ipt.maxBandwidth
}

// MaxRangeMeters returns maximum range.
func (ipt *InProcessTransport) MaxRangeMeters() int32 {
	return ipt.maxRange
}

// PowerCostRelative returns relative power cost.
func (ipt *InProcessTransport) PowerCostRelative() int32 {
	return ipt.powerCost
}

// MaxConcurrentPeers returns max concurrent peers.
func (ipt *InProcessTransport) MaxConcurrentPeers() int32 {
	return ipt.maxConcurrency
}

// SendAsync sends data to a peer asynchronously.
func (ipt *InProcessTransport) SendAsync(ctx context.Context, peerUhid string, data []byte) (bool, error) {
	if !ipt.available {
		return false, fmt.Errorf("transport not available")
	}

	if len(peerUhid) == 0 {
		return false, fmt.Errorf("peer UHID cannot be empty")
	}

	if len(data) == 0 {
		return false, fmt.Errorf("data cannot be empty")
	}

	// Check if peer exists and has a message handler
	handlerInterface, ok := ipt.messageHandlers.Load(peerUhid)
	if !ok {
		return false, fmt.Errorf("peer %s not registered", peerUhid)
	}

	handler, ok := handlerInterface.(chan []byte)
	if !ok {
		return false, fmt.Errorf("invalid handler for peer %s", peerUhid)
	}

	// Send data to the peer's channel (non-blocking)
	select {
	case handler <- append([]byte{}, data...):
		ipt.m.RecordSample(0, true, int64(len(data)))
		return true, nil
	case <-ctx.Done():
		return false, ctx.Err()
	default:
		return false, fmt.Errorf("peer handler queue full for %s", peerUhid)
	}
}

// SendStreamAsync sends stream data to a peer (equivalent to SendAsync for in-process).
func (ipt *InProcessTransport) SendStreamAsync(ctx context.Context, peerUhid string, data []byte) (bool, error) {
	return ipt.SendAsync(ctx, peerUhid, data)
}

// IsConnected checks if a peer is registered in this transport.
func (ipt *InProcessTransport) IsConnected(peerUhid string) bool {
	_, exists := ipt.connectedPeers.Load(peerUhid)
	return exists
}

// RegisterPeer registers a peer in the transport and returns a receive channel.
func (ipt *InProcessTransport) RegisterPeer(peerUhid string, bufferSize int) (chan []byte, error) {
	if len(peerUhid) == 0 {
		return nil, fmt.Errorf("peer UHID cannot be empty")
	}

	handler := make(chan []byte, bufferSize)
	ipt.messageHandlers.Store(peerUhid, handler)
	ipt.connectedPeers.Store(peerUhid, true)

	return handler, nil
}

// UnregisterPeer unregisters a peer from the transport.
func (ipt *InProcessTransport) UnregisterPeer(peerUhid string) {
	handlerInterface, ok := ipt.messageHandlers.Load(peerUhid)
	if ok {
		if handler, ok := handlerInterface.(chan []byte); ok {
			close(handler)
		}
		ipt.messageHandlers.Delete(peerUhid)
	}
	ipt.connectedPeers.Delete(peerUhid)
}

// Metrics returns the per-transport EWMA metrics for this transport.
func (ipt *InProcessTransport) Metrics() *PerTransportMetrics {
	return ipt.m
}

// Shutdown cleanly shuts down the transport.
func (ipt *InProcessTransport) Shutdown() error {
	ipt.available = false
	// Unregister all peers
	ipt.messageHandlers.Range(func(key, value interface{}) bool {
		if peerUhid, ok := key.(string); ok {
			ipt.UnregisterPeer(peerUhid)
		}
		return true
	})
	return nil
}
